custom/ewsghana/handlers/alerts.py

```diff
@@ -1,11 +1,63 @@
+from collections import defaultdict
+from casexml.apps.stock.const import SECTION_TYPE_STOCK
+from casexml.apps.stock.models import StockTransaction
+from corehq.apps.locations.models import SQLLocation
+from corehq.apps.products.models import SQLProduct
+from corehq.toggles import EWS_INVALID_REPORT_RESPONSE
+from custom.ewsghana.reminders import ERROR_MESSAGE
 from custom.ilsgateway.tanzania.handlers.keyword import KeywordHandler
-from custom.ewsghana.alerts.alerts import report_completion_check, stock_alerts
+from custom.ewsghana.alerts.alerts import stock_alerts
 from corehq.apps.sms.api import send_sms_to_verified_number
 from corehq.apps.commtrack.sms import *
 
 
+def get_transactions_by_product(transactions):
+    result = defaultdict(list)
+    for tx in transactions:
+        result[tx.product_id].append(tx)
+    return result
+
+
 class AlertsHandler(KeywordHandler):
 
+    def get_valid_reports(self, data, verified_contact):
+        filtered_transactions = []
+        excluded_products = []
+        for product_id, transactions in get_transactions_by_product(data['transactions']).iteritems():
+            begin_soh = None
+            end_soh = None
+            receipt = 0
+            for transaction in transactions:
+                if begin_soh is None:
+                    sql_location = SQLLocation.objects.get(location_id=transaction.location_id)
```

Comment:
are locations ever shared across transactions in the same message? if so this could be cached/optimized